plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/GoogleApiFactory.java

```diff
@@ -26,19 +28,48 @@
 import com.google.api.services.cloudresourcemanager.CloudResourceManager;
 import com.google.api.services.cloudresourcemanager.CloudResourceManager.Projects;
 import com.google.cloud.tools.eclipse.util.CloudToolsInfo;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URL;
 
 /**
  * Class to obtain various Google cloud Platform related APIs.
+ *
+ * TODO move this class into a separate API bundle
+ * https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1438
  */
 public class GoogleApiFactory {
 
-  private final JsonFactory jsonFactory = new JacksonFactory();
-  private final HttpTransport transport = new NetHttpTransport();
+  private static final class TimeoutAwareConnectionFactory extends DefaultConnectionFactory {
+    private final int connectTimeout;
+    private final int readTimeout;
+
+    private TimeoutAwareConnectionFactory(int connectTimeout, int readTimeout) {
+      super();
+      this.connectTimeout = connectTimeout;
+      this.readTimeout = readTimeout;
+    }
+
+    @Override
+    public HttpURLConnection openConnection(URL url) throws IOException {
+      HttpURLConnection connection = super.openConnection(url);
+      connection.setConnectTimeout(connectTimeout);
+      connection.setReadTimeout(readTimeout);
+      return connection;
+    }
+  }
+
+  private static final int DEFAULT_TIMEOUT_MS = 10000;
```

Comment:
We should probably reduce this to one second until we've done that then, and definitely not block deploy if it doesn't return fast enough.